core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -261,7 +258,20 @@ public void close() {
 
     @Override
     protected boolean shouldKeep(T posDelete) {
-      return CHARSEQ_COMPARATOR.compare(dataLocation, (CharSequence) FILENAME_ACCESSOR.get(posDelete)) == 0;
+      return charSeqEquals(dataLocation, (CharSequence) FILENAME_ACCESSOR.get(posDelete));
+    }
+
+    private boolean charSeqEquals(CharSequence s1, CharSequence s2) {
+      if (s1 == s2) {
+        return true;
+      }
+
+      int count = s1.length();
+      if (count != s2.length()) {
+        return false;
+      }
+
+      return s1.toString().equals(s2.toString());
```

Comment:
only need to call "toString" on one of these since String has a contentsEquals(CharSequence) method

```
boolean | contentEquals(CharSequence cs)Compares this string to the specified CharSequence.
```